File: 6_kyu/Reverse_or_rotate.go
package kata

// Check if sum of all digits square is even
func isSumDiv2(chunk []rune) bool {
	sum := 0
	for _, r := range chunk {
		sum += int(r-'0') % 2
	}
	return sum%2 == 0
}

// Reverse the content of the slice in place.
func reverseInPlace(chunk []rune) {
	last := len(chunk) - 1
	for index := 0; index <= last/2; index++ {
		chunk[index], chunk[last-index] = chunk[last-index], chunk[index]
	}
}

// Rotate left the content of the slice one element.
func rotateInPlace(chunk []rune) {
	first := chunk[0]
	for index := 1; index < len(chunk); index++ {
		chunk[index-1] = chunk[index]
	}
	chunk[len(chunk)-1] = first
}

func Revrot(s string, n int) string {
	if n <= 0 || len(s) < n {
		return ""
	} else {
		runes := []rune(s)
		nrChunks := len(s) / n
		for startIndex := 0; startIndex+n <= len(s); startIndex += n {
			chunk := runes[startIndex : startIndex+n]
			if isSumDiv2(chunk) {
				reverseInPlace(chunk)
			} else {
				rotateInPlace(chunk)
			}
		}
		return string(runes[:nrChunks*n])
	}
}

File: 6_kyu/Reverse_or_rotate.lua
local solution = {}

-- Reverses or rotates string s.
-- Reverses the string if the sum of all digits square is even.
local function revRot(s)
  local sum = 0
  for i = 1,#s do
    -- x is even if and only x² is even.
    -- digit value of x is even if ASCII value of x is even.
    sum = (sum + s:sub(i,i):byte()) % 2
  end
  if sum == 0 then
    return s:reverse()
  else
    return s:sub(2) .. s:sub(1,1)
  end
end

-- Splits string s into chunks of chunkSize length, reverse or rotates the chunks.
function solution.revRot(s, chunkSize)
  if #s < chunkSize or chunkSize < 1 then
    return ""
  end
  local result = ""
  for start = 1,chunkSize*(#s//chunkSize),chunkSize do
    local chunk = s:sub(start,start+chunkSize-1)
    result = result .. revRot(chunk)
  end
  return result
end
  
return solution
